Properly render HRTBs

Stupremee opened the pull request:
```rust
pub fn test<T>() 
where
    for<'a> &'a T: Iterator,
{}
```

This will now render properly including the `for<'a>`
![image](https://user-images.githubusercontent.com/39732259/116808426-fe6ce600-ab38-11eb-9452-f33f554fbb8e.png)

I do not know if this covers all cases, it only covers everything that I could think of that includes `for` and lifetimes in where bounds.
Also someone need to mentor me on how to add a proper rustdoc test for this.

Resolves #78482

rust-highfive commented:
r? @jyn514

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2485889531) [(plain)](https://github.com/rust-lang/rust/commit/c7343094495729f920c2239b9f886ef963e53910/checks/2485889531/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11813
.................................................................................................... 9500/11813
.................................................................................................... 9600/11813
...................................i......i......................................................... 9700/11813
.................................................................................iiiiiii..iiiiii.i.. 9800/11813
.................................................................................................... 10000/11813
.................................................................................................... 10100/11813
.................................................................................................... 10200/11813
.................................................................................................... 10300/11813
---
 finished in 0.410 seconds
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 33 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiii....
test result: ok. 4 passed; 0 failed; 29 ignored; 0 measured; 0 filtered out; finished in 0.11s

 finished in 0.175 seconds
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii..ii..i..i..ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.27s

 finished in 2.334 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
failures:

---- [rustdoc] rustdoc/synthetic_auto/complex.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/synthetic_auto/complex" "/checkout/src/test/rustdoc/synthetic_auto/complex.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
23: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//*[@id="synthetic-implementations-list"]//*[@class="impl"]//code' "impl<'a, T, K: ?Sized> Send for Outer<'a, T, K> where K: for<'b> Fn((&'b bool, &'a u8)) -> &'b i8, T: MyTrait<'a>, <T as MyTrait<'a>>::MyItem: Copy, 'a: 'static"
Encountered 1 errors

------------------------------------------

---
test result: FAILED. 433 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 30.57s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:14:17
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2486120933) [(plain)](https://github.com/rust-lang/rust/commit/a01f4d7c5bf5d31b735cd41bcc705ba1d7dc4654/checks/2486120933/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11813
.................................................................................................... 9500/11813
.................................................................................................... 9600/11813
...................................i......i......................................................... 9700/11813
.................................................................................iiiiiii..iiiiii.i.. 9800/11813
.................................................................................................... 10000/11813
.................................................................................................... 10100/11813
.................................................................................................... 10200/11813
.................................................................................................... 10300/11813
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 33 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiii....

 finished in 0.183 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i...i.ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.52s

 finished in 2.593 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
failures:

---- [rustdoc] rustdoc/synthetic_auto/complex.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/synthetic_auto/complex" "/checkout/src/test/rustdoc/synthetic_auto/complex.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
23: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//*[@id="synthetic-implementations-list"]//*[@class="impl"]//code' "impl<'a, T, K: ?Sized> Send for Outer<'a, T, K> where K: for<'b> Fn((&'b bool, &'a u8)) -> &'b i8, T: MyTrait<'a>, <T as MyTrait<'a>>::MyItem: Copy, 'a: 'static"
Encountered 1 errors

------------------------------------------

---
test result: FAILED. 433 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 32.17s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:15:40
```

</details>

camelid commented:
>  Also someone need to mentor me on how to add a proper rustdoc test for this.

It looks like you already added a test? But could you also add a test case for HRTB in struct and enum definitions? For example:

```rust
struct Foo<'b> {
    pub my_fn: for<'a> fn(&'a u32) -> &'b u32,
}

// and something equivalent for enums
```

jyn514 commented:
@camelid are you interested in reviewing this? I have a lot of PRs assigned :sweat_smile: and I'm about to go on break for a few weeks.

Stupremee commented:
> @camelid are you interested in reviewing this? I have a lot of PRs assigned :sweat_smile: and I'm about to go on break for a few weeks.

I will stop assigning you! Have fun!

camelid commented:
> @camelid are you interested in reviewing this? I have a lot of PRs assigned :sweat_smile: and I'm about to go on break for a few weeks.

Sure! Enjoy your break :)

r? @camelid

camelid commented on src/test/rustdoc/higher-ranked-trait-bounds.rs:
```diff
@@ -0,0 +1,41 @@
+#![crate_name = "foo"]
+
```
Here's another case where `for<'a>` can show up; could you test this as well?

```rust
impl A<'a> for dyn for<'a> A<'a> {}
```

camelid replied:
You may want to look at `rustc_hir_pretty`'s implementation of printing HRTB for ideas. You should be able to find the different spots by searching for `"for` (e.g., `rg '"for' compiler/rustc_hir_pretty`).

Stupremee replied:
oh wow, didn't even knew this is all valid code :smile: 

Will add cases for that!

Stupremee replied:
Okay it does not work yet, but I figured out a solution!

Stupremee commented:
I need your opionion on this @camelid. The "best" solution I found, that is not just a dirty fix, would be to introduce a new `Type` called `TraitObject` which would contain all important data. However, that seems like a big change which might need more discussion? Since this could also influence rustdoc-json and probably more.

camelid commented:
> The "best" solution I found, that is not just a dirty fix, would be to introduce a new `Type` called `TraitObject` which would contain all important data.

Hmm, but how would that handle the case of bare functions like `for<'a> fn(&'a i32) -> i32`? And also how would it work with `where` bounds?

I tried looking in the Rust Reference to see what other places HRTBs can be used, but unfortunately it only seems to include information about uses in `where` clauses, and not bare functions.

I also found another place where HRTBs can show up, though it probably is handled the same way as with `where` in `impl`s:

```rust
struct Foo<T>
where
    for<'a> T: PartialEq<&'a T>,
{
    x: T,
}

struct Bar<T>
where
    T: for<'a> PartialEq<&'a T>,
{
    x: T,
}
```

I would recommend asking on Zulip for help with figuring out where HRTBs can appear and how they should be handled. Hopefully some people more knowledgeable about the intricacies of HRTB can help you there :)

Stupremee commented:
> Hmm, but how would that handle the case of bare functions like for<'a> fn(&'a i32) -> i32? And also how would it work with where bounds?

These are both already handled. My current fix includes a new `Type` because the information which lifetimes are specified was previously discarded while cleaning, so I needed to introduce a new `Type` called `PolyTraitTy` (maybe `TraitObject` is a better name, not sure). 
I also removed the old `for-lifetime.rs` test and moved it into `higher-kind-trait-bounds.rs` together with some other tests.

Stupremee commented:
r? @camelid 
This should be ready now!

camelid reviewed (commented):
Thank you for working on this!

This is a fairly tricky bug, so I left some fairly detailed comments. Let me know if you want me to clarify anything :)

Stupremee commented on src/librustdoc/clean/types.rs:
```diff
@@ -1460,7 +1460,9 @@ crate enum Type {
     ResolvedPath {
         path: Path,
-        param_names: Option<Vec<GenericBound>>,
+        /// If `param_names` is `Some`, this path is a trait object and the Vecs repsresent
+        /// `(generic bounds, generic parameters)`
+        param_names: Option<(Vec<GenericBound>, Vec<GenericParamDef>)>,
```
I somehow expected to get questions on the code because It's *really* weird, I know :sweat_smile: 

> Also, it seems odd to me that data specific to dyn Traits is stored as part of ResolvedPath. I noticed that you said in #84814 (comment) that you were planning to add a new Type variant but decided to instead change the existing ResolvedPath field. I think your original idea would actually be a better approach, and would make the code easier to understand.

Yes I agree that this is definitely cleaner and easier to understand. I will change that!

> I feel a bit confused about this. PathSegments (Path is mostly a Vec<PathSegment>) have generic arguments in them, so why are generic arguments both there and here?

This is the "hack" that rustdoc currently uses. It uses the `Path` to display trait objects. Before this PR, the `param_names` field would indicate if this `Path` is a trait object. If the `param_names` is `Some`, then this `Path` is a trait object and it's bounds are *this* path, and all the bounds in the `param_names` (`param_names` is basically a list of [these](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.PolyTraitRef.html#structfield.trait_ref) elements, where the first trait ref is special-cased and becomes *this* path) field. The second list in the tuple, was a list containing [these](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.PolyTraitRef.html#structfield.bound_generic_params) elements, so it includes all lifetimes that were provided using `for<...>`.

Stupremee replied:
> Could you explain a bit why you added the new field?

The `bounds` part does not include the higher ranked lifetimes. If you look at the doc for the rustic [WhereBoundPredicate](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.WhereBoundPredicate.html), you'll see that there's a `bounds` field, which directly translates to the `bounds` field in the cleaned `WherePredicate`, and a `bound_generic_params`, which includes all lifetimes provided by `for<>` (as written in the documentation). The latter field was previously discarded while cleaning the predicate, so I added a field to keep the information about these lifetimes.

Stupremee commented:
@camelid Sorry for the long delay here.

My new commit introduces the new `DynTrait` type, and in general, everything works really well. 
However, there are still two tests failing, due to cross crate re-exports, but I wasn't able to find out why they are failing, yet.

Might you be able to look at them and see if you can spot the mistake faster than me spending hours debbuing :sweat_smile: ?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2861424968) [(plain)](https://github.com/rust-lang/rust/commit/3f9f7db31efa949edf84ba70a5bae903728ba6a2/checks/2861424968/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error: unused import: `MacroKind`
 --> src/librustdoc/clean/utils.rs:6:5
  |
6 |     MacroKind, PolyTrait,
  |
  |
  = note: `-D unused-imports` implied by `-D warnings`
error: aborting due to previous error

error: could not compile `rustdoc`

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2861470911) [(plain)](https://github.com/rust-lang/rust/commit/f90d268afdb0c8d0f544e632a52ddbb3a16fad77/checks/2861470911/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Some tests failed in compiletest suite=rustdoc mode=rustdoc host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu

---- [rustdoc] rustdoc/inline_cross/issue-32881.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/inline_cross/issue-32881" "/checkout/src/test/rustdoc/inline_cross/issue-32881.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
8: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//code' "impl<'a> dyn Bar"
9: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//code' "impl<'a> Debug for dyn Bar"
Encountered 2 errors

------------------------------------------

---
test result: FAILED. 442 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 29.99s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:14:50
```

</details>

camelid commented:
> However, there are still two tests failing, due to cross crate re-exports, but I wasn't able to find out why they are failing, yet.
> 
> Might you be able to look at them and see if you can spot the mistake faster than me spending hours debbuing 😅 ?

Could you post screenshots of what the generated pages look like for the parts of the tests that are failing? I probably won't be able to look at them for the next week or so, but I'll make sure to take a look after.

Stupremee commented:
Ahh, I finally found it. This should be ready to merge now!

r? @camelid

GuillaumeGomez commented:
Thanks!

@bors: r+

bors commented:
:pushpin: Commit 4ea27484c9582389cded2301d1f0dd4f421c8c35 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez 4ea27484c9582389cded2301d1f0dd4f421c8c35 -->
<!-- homu: {"type":"Approved","sha":"4ea27484c9582389cded2301d1f0dd4f421c8c35","approver":"GuillaumeGomez"} -->

bors commented:
:hourglass: Testing commit 4ea27484c9582389cded2301d1f0dd4f421c8c35 with merge 831ae3c1364b7b033bd1da430bc1cb86d93ad186...
<!-- homu: {"type":"BuildStarted","head_sha":"4ea27484c9582389cded2301d1f0dd4f421c8c35","merge_sha":"831ae3c1364b7b033bd1da430bc1cb86d93ad186"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2921151446)
Approved by: GuillaumeGomez
Pushing 831ae3c1364b7b033bd1da430bc1cb86d93ad186 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"GuillaumeGomez","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2921151446"},"merge_sha":"831ae3c1364b7b033bd1da430bc1cb86d93ad186"} -->
